### PR TITLE
Adds markdown parser context

### LINKED ISSUE #284: Any way of accessing (external) contextual data in a parser?
I've written a MarkDig parser and in the parser I would want to get access to some contextual information about the document being parsed (that's not available in the document itself).

Is there any way of passing contextual data to a parser on a parse basis (just like I can in Scriban when rendering)?

```csharp
// Something like this?
var html = Markdown.ToHtml(markdown, _pipeline, context);
```

If there is no work around for my problem, I would be happy to provide a pull request for this functionality if you think it would be a good fit for MarkDig.

### PR DESCRIPTION
**Pull request by patriksvensson**
Closes #284

### CODE REVIEW

**Comment by patriksvensson** [comment]
I've added the new context parameter to the methods I thought made sense in `Markdown.cs`, but would want some feedback on whether or not this is correct or not.

File: src/Markdig/MarkdownParserContext.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+{
+    /// <summary>
+    /// Provides a context that can be used as part of parsing Markdown documents.
+    /// </summary>
+    public sealed class MarkdownParserContext
+    {
+        /// <summary>
+        /// Gets or sets the context property collection.
+        /// </summary>
+        public IDictionary<object, object> Properties { get; set; }
```
Comment (xoofx) [inline src/Markdig/MarkdownParserContext.cs]: You can expose only a `Dictionary<object, object>`, not a `IDictionary` and remove also the `setter`.

Comment (xoofx) [reply src/Markdig/MarkdownParserContext.cs]: Nevermind, forgot that I can do that right away from github, so simple 😅

**Comment by xoofx** [comment]
Thanks!

**Comment by sveins12** [comment]
How can you access the Properties dictionary from within a markdown extension?

**Pull request merged**